torch/testing/_internal/common_utils.py

```diff
@@ -832,219 +939,241 @@ def safeCoalesce(self, t):
-
-        a_tol = self.get_default_tolerance(a)
-        b_tol = self.get_default_tolerance(b)
-        return (max(a_tol[0], b_tol[0]), max(a_tol[1], b_tol[1]))
-
-    def assertEqual(self, x, y, message='', *, atol=None, rtol=None, allow_inf=False, exact_dtype=None):
+    # Checks if two dense tensors are equal(-ish), returning (True, None)
+    # when they are and (False, debug_msg) when they are not, where the
+    # debug_msg is a string containing information about why the tensors
+    # failed to compare as equal(-ish).
+    # If exact_dtype is true both tensors must have the same dtype.
+    # If exact_device is true both tensors must be on the same device.
+    # See NOTE Test Framework Tensor "Equality" for more information on
+    # how tensors are compared.
+    # NOTE: tensors on different devices are moved to the CPU to be compared when
+    # exact_device is False.
+    def _compareTensors(self, a, b, *, rtol=None, atol=None, equal_nan=True,
+                        exact_dtype=False, exact_device=False):
+        if not isinstance(a, torch.Tensor):
+            return (False, "argument a, {0}, to _compareTensors is not a tensor!".format(a))
+        if not isinstance(b, torch.Tensor):
+            return (False, "argument b, {0}, to _compareTensors is not a tensor!".format(b))
+
+        # Validates tensors are on the same device
+        if exact_device and a.device != b.device:
+            return (False, ("Attempted to compare equality of tensors on "
+                            "different devices! Got devices {0} and "
+                            "{1}.".format(a.device, b.device)))
+
+        # Compares tensors of different devices on the CPU
+        if a.device != b.device:
+            a = a.cpu()
+            b = b.cpu()
+
+        # Checks size matches
+        if a.size() != b.size():
+            return (False, ("Attempted to compare equality of tensors with "
+                            "different sizes. Got sizes {0} and {1}.").format(a.size(), b.size()))
+
+        # Checks dtype (if exact_dtype)
+        if exact_dtype and a.dtype is not b.dtype:
+            return (False, ("Attempted to compare equality of tensors with "
+                            "different dtypes. Got dtypes {0} and {1}.").format(a.dtype, b.dtype))
+
+        # Acquires atol and rtol
+        # NOTE: atol can be overriden by a class's precision value. This
+        # legacy behavior allows generated tests to customize their precision.
+        rtol = rtol if rtol is not None else max(self.dtype_precisions.get(a.dtype, (0, 0))[0],
+                                                 self.dtype_precisions.get(b.dtype, (0, 0))[0])
+        atol = atol if atol is not None else max(self.dtype_precisions.get(a.dtype, (0, 0))[1],
+                                                 self.dtype_precisions.get(b.dtype, (0, 0))[1])
+        atol = max(atol, self.precision)
```

Comment:
on master, when we honor self.precision we only use atol. This PR seems like it would add an rtol on tests that currently only check an atol (loosening the tolerance).

Reply:
Master's behavior is actually a little wonky in this case and doesn't respect the intent of self.precision, I believe this new PR corrects the behavior.

One of the uses of self.precision is for generated tests to set a more lenient atol, which is what happens here. This PR also updates the default self.precision to be zero so unless it's explicitly set this max will be a nullop.

Current master seems to use self.precision as a fallback value for atol and rtol, which was not the intent. I also don't think setting self.precision would ever override the given atol value, but self.precision is explicitly intended to override atol.

Reply:
>Current master seems to use self.precision as a fallback value for atol and rtol

where do you see that? I believe I updated uses of self.precision to `atol=self.precision`.

Reply:
I thought so, too, but on current master here's all I can find: https://github.com/pytorch/pytorch/blob/dc2519083380efe02e4e813536714581b3f8f396/torch/testing/_internal/common_utils.py#L867.

Reply:
That's probably not ideal (and maybe unnecessary), but it only applies when no dtype is found. This PR would apply a potentially-higher rtol even when an rtol is found for that dtype.

Reply:
You mean a higher atol? Yes, that is the intent of self.precision. It is to allow overriding the atol.

Reply:
I mean that where required (and with the intent to remove the attribute later) I updated tests depending on self.precision to call `self.assertEqual(x, y, atol=self.precision)`. Explicit atol sets rtol to zero. If I missed cases (which is likely), this would set atol=self.precision _and_ set rtol to the default, which would widen the total tolerance.

Reply:
Oh I see. Yes that's true that before when atol was explicitly set then rtol was set to zero. I think the current behavior is more consistent with isclose, however, which does not zero rtol if only atol is set.

Reply:
> isclose, however, which does not zero rtol if only atol is set.

But we're preserving the behavior of setting rtol=0 when only an atol is passed to assertEqual, right?

Reply:
Only in the legacy case where atol is specified positionally instead of "message." In all the non-legacy cases they're set independently.

I suppose that's an odd legacy case to not update, though? I originally thought it was analogous to "deprecated" behavior, but it seems like we should make it consistent, too?